File: frontend/src/features/bookings/services/bookingService.ts
import axios from 'axios';
import authService from '../../auth/services/authService';
import type { 
  Booking, 
  BookingListResponse, 
  BookingPaginationParams, 
  CreateBookingRequest, 
  AvailableRoomRequest,
  CheckoutSummary
} from '../types';
import type { Room } from '../../rooms/types';

const API_BASE_URL = 'http://localhost:8080/api';

const api = axios.create({
  baseURL: API_BASE_URL,
  headers: {
    'Content-Type': 'application/json',
  },
});

// Add token to requests
api.interceptors.request.use(async (config) => {
  try {
    const token = await authService.getValidToken();
    if (token) {
      config.headers.Authorization = `Bearer ${token}`;
    }
  } catch (error) {
    console.error('Failed to get valid token:', error);
  }
  return config;
});

// Add response interceptor for error handling
api.interceptors.response.use(
  (response) => response,
  async (error) => {
    if (error.response?.status === 401) {
      console.log('Authentication failed - attempting token refresh');
      try {
        const newToken = await authService.getValidToken();
        if (newToken && error.config) {
          // Retry the original request with new token
          error.config.headers.Authorization = `Bearer ${newToken}`;
          return api.request(error.config);
        }
      } catch (refreshError) {
        console.error('Token refresh failed:', refreshError);
        authService.clearTokens();
        window.location.href = '/';
      }
    }
    return Promise.reject(error);
  }
);

export const bookingService = {
  /**
   * Get all bookings with pagination and filtering
   */
  async getAllBookings(params: BookingPaginationParams): Promise<BookingListResponse> {
    try {
      const queryParams = new URLSearchParams({
        page: (params.page - 1).toString(), // Backend uses 0-based indexing
        size: params.size.toString(),
        ...(params.sortBy && { sortBy: params.sortBy }),
        ...(params.sortDirection && { sortDirection: params.sortDirection }),
        ...(params.filters?.search && { search: params.filters.search }),
        ...(params.filters?.status && { status: params.filters.status }),
      });

      const response = await api.get(`/bookings?${queryParams}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching bookings:', error);
      throw error;
    }
  },

  /**
   * Get booking by ID
   */
  async getBookingById(id: number): Promise<Booking> {
    try {
      const response = await api.get(`/bookings/${id}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching booking:', error);
      throw error;
    }
  },

  /**
   * Create new booking
   */
  async createBooking(booking: CreateBookingRequest): Promise<Booking> {
    try {
      const response = await api.post('/bookings', booking);
      return response.data;
    } catch (error) {
      console.error('Error creating booking:', error);
      throw error;
    }
  },

  /**
   * Update booking status
   */
  async updateBookingStatus(id: number, status: string): Promise<Booking> {
    try {
      const response = await api.put(`/bookings/${id}/status?status=${status}`);
      return response.data;
    } catch (error) {
      console.error('Error updating booking status:', error);
      throw error;
    }
  },

  /**
   * Cancel booking
   */
  async cancelBooking(id: number): Promise<Booking> {
    try {
      const response = await api.put(`/bookings/${id}/cancel`);
      return response.data;
    } catch (error) {
      console.error('Error cancelling booking:', error);
      throw error;
    }
  },

  /**
   * Check in guest
   */
  async checkInGuest(id: number): Promise<Booking> {
    try {
      const response = await api.put(`/bookings/${id}/check-in`);
      return response.data;
    } catch (error) {
      console.error('Error checking in guest:', error);
      throw error;
    }
  },

  /**
   * Check out guest
   */
  async checkOutGuest(id: number): Promise<Booking> {
    try {
      const response = await api.put(`/bookings/${id}/check-out`);
      return response.data;
    } catch (error) {
      console.error('Error checking out guest:', error);
      throw error;
    }
  },

  /**
   * Get checkout summary with billing details
   */
  async getCheckoutSummary(id: number): Promise<CheckoutSummary> {
    try {
      const response = await api.get(`/bookings/${id}/checkout-summary`);
      return response.data;
    } catch (error) {
      console.error('Error fetching checkout summary:', error);
      throw error;
    }
  },

  /**
   * Get available rooms for booking dates
   */
  async getAvailableRooms(params: AvailableRoomRequest): Promise<Room[]> {
    try {
      const queryParams = new URLSearchParams({
        checkInDate: params.checkInDate,
        checkOutDate: params.checkOutDate,
        adultCapacity: params.adultCapacity.toString(),
        childrenCapacity: params.childrenCapacity.toString(),
      });

      const response = await api.get(`/bookings/available-rooms?${queryParams}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching available rooms:', error);
      throw error;
    }
  },

  /**
   * Get upcoming check-ins
   */
  async getUpcomingCheckIns(): Promise<Booking[]> {
    try {
      const response = await api.get('/bookings/upcoming-checkins');
      return response.data;
    } catch (error) {
      console.error('Error fetching upcoming check-ins:', error);
      throw error;
    }
  },

  /**
   * Get current guests
   */
  async getCurrentGuests(): Promise<Booking[]> {
    try {
      const response = await api.get('/bookings/current-guests');
      return response.data;
    } catch (error) {
      console.error('Error fetching current guests:', error);
      throw error;
    }
  },

  /**
   * Get total booking count
   */
  async getTotalBookingCount(): Promise<number> {
    try {
      const response = await api.get('/bookings/count');
      return response.data.count;
    } catch (error) {
      console.error('Error fetching booking count:', error);
      throw error;
    }
  },

  /**
   * Get booking count by status
   */
  async getBookingCountByStatus(status: string): Promise<number> {
    try {
      const response = await api.get(`/bookings/count/${status}`);
      return response.data.count;
    } catch (error) {
      console.error('Error fetching booking count by status:', error);
      throw error;
    }
  },
};

export default bookingService;

File: backend/src/main/java/com/booking/hotel/service/BookingService.java
package com.booking.hotel.service;

import com.booking.hotel.dto.BookingRequestDTO;
import com.booking.hotel.dto.BookingResponseDTO;
import com.booking.hotel.dto.CheckInSummaryDTO;
import com.booking.hotel.dto.CheckoutSummaryDTO;
import com.booking.hotel.dto.ServiceChargeDTO;
import com.booking.hotel.entity.Booking;
import com.booking.hotel.entity.Room;
import com.booking.hotel.entity.RoomService;
import com.booking.hotel.repository.BookingRepository;
import com.booking.hotel.repository.RoomRepository;
import com.booking.hotel.repository.RoomServiceRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import jakarta.persistence.LockModeType;
import jakarta.persistence.EntityManager;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.temporal.ChronoUnit;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
@Transactional
public class BookingService {

    @Autowired
    private BookingRepository bookingRepository;

    @Autowired
    private RoomRepository roomRepository;

    @Autowired
    private EntityManager entityManager;
    
    @Autowired
    private RoomServiceRepository roomServiceRepository;

    /**
     * Get all bookings with pagination, sorting, and search
     */
    public Page<BookingResponseDTO> getAllBookings(int page, int size, String sortBy, 
                                                   String sortDirection, String search, 
                                                   Booking.BookingStatus status) {
        try {
            // Validate sort direction
            Sort.Direction direction = sortDirection.equalsIgnoreCase("desc") ? 
                Sort.Direction.DESC : Sort.Direction.ASC;

            // Create sort object
            Sort sort = Sort.by(direction, sortBy);
            Pageable pageable = PageRequest.of(page, size, sort);

            // Fetch bookings with search and status filter
            Page<Booking> bookingPage = bookingRepository.findAllWithSearch(search, status, pageable);

            // Convert to DTOs
            return bookingPage.map(BookingResponseDTO::new);
        } catch (Exception e) {
            throw new RuntimeException("Error fetching bookings: " + e.getMessage(), e);
        }
    }

    /**
     * Get booking by ID
     */
    public Optional<BookingResponseDTO> getBookingById(Long id) {
        try {
            return bookingRepository.findById(id)
                    .map(BookingResponseDTO::new);
        } catch (Exception e) {
            throw new RuntimeException("Error fetching booking with id " + id + ": " + e.getMessage(), e);
        }
    }

    /**
     * Create new booking with availability check and pessimistic locking
     */
    @Transactional
    public BookingResponseDTO createBooking(BookingRequestDTO bookingRequest) {
        try {
            // Validation
            if (!bookingRequest.isValidDateRange()) {
                throw new RuntimeException("Invalid date range: Check-out date must be after check-in date");
            }

            if (bookingRequest.getCheckInDate().isBefore(LocalDate.now())) {
                throw new RuntimeException("Check-in date cannot be in the past");
            }

            // Get room with pessimistic lock to prevent concurrent bookings
            Room room = roomRepository.findById(bookingRequest.getRoomId())
                    .orElseThrow(() -> new RuntimeException("Room not found with id: " + bookingRequest.getRoomId()));

            // Apply pessimistic lock
            entityManager.lock(room, LockModeType.PESSIMISTIC_WRITE);

            // Check room capacity
            if (room.getAdultCapacity() < bookingRequest.getAdultCapacity()) {
                throw new RuntimeException("Room adult capacity (" + room.getAdultCapacity() + 
                        ") is insufficient for requested capacity (" + bookingRequest.getAdultCapacity() + ")");
            }

            if (room.getChildrenCapacity() < bookingRequest.getChildrenCapacity()) {
                throw new RuntimeException("Room children capacity (" + room.getChildrenCapacity() + 
                        ") is insufficient for requested capacity (" + bookingRequest.getChildrenCapacity() + ")");
            }

            // Check availability
            boolean isAvailable = bookingRepository.isRoomAvailable(
                    room, 
                    bookingRequest.getCheckInDate(), 
                    bookingRequest.getCheckOutDate()
            );

            if (!isAvailable) {
                throw new RuntimeException("Room " + room.getRoomNumber() + 
                        " is not available for the selected dates");
            }

            // Calculate total amount
            long numberOfNights = ChronoUnit.DAYS.between(
                    bookingRequest.getCheckInDate(), 
                    bookingRequest.getCheckOutDate()
            );
            BigDecimal totalAmount = room.getPrice().multiply(BigDecimal.valueOf(numberOfNights));

            // Create booking entity
            Booking booking = new Booking(
                    bookingRequest.getFirstName(),
                    bookingRequest.getLastName(),
                    bookingRequest.getPronouns(),
                    bookingRequest.getCheckInDate(),
                    bookingRequest.getCheckOutDate(),
                    bookingRequest.getAdultCapacity(),
                    bookingRequest.getChildrenCapacity(),
                    room,
                    totalAmount
            );

            // Save booking
            Booking savedBooking = bookingRepository.save(booking);

            return new BookingResponseDTO(savedBooking);
        } catch (RuntimeException e) {
            throw e;
        } catch (Exception e) {
            throw new RuntimeException("Error creating booking: " + e.getMessage(), e);
        }
    }

    /**
     * Update booking status
     */
    public BookingResponseDTO updateBookingStatus(Long id, Booking.BookingStatus status) {
        try {
            Booking booking = bookingRepository.findById(id)
                    .orElseThrow(() -> new RuntimeException("Booking not found with id: " + id));

            booking.setStatus(status);
            Booking updatedBooking = bookingRepository.save(booking);

            return new BookingResponseDTO(updatedBooking);
        } catch (RuntimeException e) {
            throw e;
        } catch (Exception e) {
            throw new RuntimeException("Error updating booking status: " + e.getMessage(), e);
        }
    }

    /**
     * Cancel booking
     */
    public BookingResponseDTO cancelBooking(Long id) {
        return updateBookingStatus(id, Booking.BookingStatus.CANCELLED);
    }

    /**
     * Check in guest
     */
    public BookingResponseDTO checkInGuest(Long id) {
        try {
            Booking booking = bookingRepository.findById(id)
                    .orElseThrow(() -> new RuntimeException("Booking not found with id: " + id));

            if (booking.getStatus() != Booking.BookingStatus.BOOKED) {
                throw new RuntimeException("Cannot check in: Booking status is " + booking.getStatus());
            }

            if (booking.getCheckInDate().isAfter(LocalDate.now())) {
                throw new RuntimeException("Cannot check in before check-in date");
            }

            return updateBookingStatus(id, Booking.BookingStatus.CHECKED_IN);
        } catch (RuntimeException e) {
            throw e;
        } catch (Exception e) {
            throw new RuntimeException("Error checking in guest: " + e.getMessage(), e);
        }
    }

    /**
     * Check out guest
     */
    public BookingResponseDTO checkOutGuest(Long id) {
        try {
            Booking booking = bookingRepository.findById(id)
                    .orElseThrow(() -> new RuntimeException("Booking not found with id: " + id));

            if (booking.getStatus() != Booking.BookingStatus.CHECKED_IN) {
                throw new RuntimeException("Cannot check out: Guest is not checked in");
            }

            return updateBookingStatus(id, Booking.BookingStatus.CHECKED_OUT);
        } catch (RuntimeException e) {
            throw e;
        } catch (Exception e) {
            throw new RuntimeException("Error checking out guest: " + e.getMessage(), e);
        }
    }

    /**
     * Get available rooms for given criteria
     */
    public List<Room> getAvailableRooms(LocalDate checkInDate, LocalDate checkOutDate, 
                                       Integer adultCapacity, Integer childrenCapacity) {
        try {
            if (checkInDate == null || checkOutDate == null) {
                throw new RuntimeException("Check-in and check-out dates are required");
            }

            if (checkInDate.isAfter(checkOutDate) || checkInDate.isEqual(checkOutDate)) {
                throw new RuntimeException("Check-out date must be after check-in date");
            }

            if (checkInDate.isBefore(LocalDate.now())) {
                throw new RuntimeException("Check-in date cannot be in the past");
            }

            return bookingRepository.findAvailableRooms(
                    checkInDate, checkOutDate, 
                    adultCapacity != null ? adultCapacity : 1, 
                    childrenCapacity != null ? childrenCapacity : 0
            );
        } catch (RuntimeException e) {
            throw e;
        } catch (Exception e) {
            throw new RuntimeException("Error fetching available rooms: " + e.getMessage(), e);
        }
    }

    /**
     * Get bookings by date range
     */
    public List<BookingResponseDTO> getBookingsByDateRange(LocalDate startDate, LocalDate endDate) {
        try {
            List<Booking> bookings = bookingRepository.findByDateRange(startDate, endDate);
            return bookings.stream()
                    .map(BookingResponseDTO::new)
                    .toList();
        } catch (Exception e) {
            throw new RuntimeException("Error fetching bookings by date range: " + e.getMessage(), e);
        }
    }

    /**
     * Get upcoming check-ins
     */
    public List<BookingResponseDTO> getUpcomingCheckIns() {
        try {
            List<Booking> bookings = bookingRepository.findUpcomingCheckIns(LocalDate.now());
            return bookings.stream()
                    .map(BookingResponseDTO::new)
                    .toList();
        } catch (Exception e) {
            throw new RuntimeException("Error fetching upcoming check-ins: " + e.getMessage(), e);
        }
    }

    /**
     * Get current guests (checked in)
     */
    public List<BookingResponseDTO> getCurrentGuests() {
        try {
            List<Booking> bookings = bookingRepository.findCurrentGuests();
            return bookings.stream()
                    .map(BookingResponseDTO::new)
                    .toList();
        } catch (Exception e) {
            throw new RuntimeException("Error fetching current guests: " + e.getMessage(), e);
        }
    }

    /**
     * Get total booking count
     */
    public long getTotalBookingCount() {
        try {
            return bookingRepository.count();
        } catch (Exception e) {
            throw new RuntimeException("Error counting bookings: " + e.getMessage(), e);
        }
    }

    /**
     * Get booking count by status
     */
    public long getBookingCountByStatus(Booking.BookingStatus status) {
        try {
            return bookingRepository.countByStatus(status);
        } catch (Exception e) {
            throw new RuntimeException("Error counting bookings by status: " + e.getMessage(), e);
        }
    }

    /**
     * Get checkout summary with billing calculation
     */
    public CheckoutSummaryDTO getCheckoutSummary(Long id) {
        try {
            Booking booking = bookingRepository.findById(id)
                    .orElseThrow(() -> new RuntimeException("Booking not found with id: " + id));

            if (booking.getStatus() != Booking.BookingStatus.CHECKED_IN) {
                throw new RuntimeException("Cannot get checkout summary: Guest is not checked in");
            }

            LocalDate checkInDate = booking.getCheckInDate();
            LocalDate checkOutDate = LocalDate.now(); // Current date for checkout
            
            // Calculate total nights
            long totalNights = ChronoUnit.DAYS.between(checkInDate, checkOutDate);
            if (totalNights < 1) {
                totalNights = 1; // Minimum 1 night
            }

            // Calculate room total
            BigDecimal roomTotal = booking.getRoom().getPrice().multiply(BigDecimal.valueOf(totalNights));

            // Calculate service charges (tax + cleaning fee + room services)
            List<ServiceChargeDTO> serviceCharges = new ArrayList<>();
            
            // Add tax (10% of room total)
            BigDecimal tax = roomTotal.multiply(BigDecimal.valueOf(0.10));
            serviceCharges.add(new ServiceChargeDTO("Tax (10%)", tax));
            
            // Add cleaning fee
            BigDecimal cleaningFee = BigDecimal.valueOf(3000); // Fixed cleaning fee
            serviceCharges.add(new ServiceChargeDTO("Cleaning Fee", cleaningFee));
            
            // Add room service charges
            BigDecimal roomServicesTotal = BigDecimal.ZERO;
            List<RoomService> roomServicesList = roomServiceRepository.findByRoomNumberOrderByRequestedAtDesc(booking.getRoom().getRoomNumber());
            
            for (RoomService roomService : roomServicesList) {
                // Only include completed services
                if (roomService.getStatus() == RoomService.ServiceStatus.COMPLETED) {
                    String serviceName = roomService.getServiceType() + " - " + 
                                      (roomService.getDescription() != null ? roomService.getDescription() : "Room Service");
                    serviceCharges.add(new ServiceChargeDTO(serviceName, roomService.getAmount()));
                    roomServicesTotal = roomServicesTotal.add(roomService.getAmount());
                }
            }

            // Calculate grand total
            BigDecimal grandTotal = roomTotal.add(tax).add(cleaningFee).add(roomServicesTotal);

            return new CheckoutSummaryDTO(
                roomTotal,
                serviceCharges,
                grandTotal,
                (int) totalNights,
                checkInDate,
                checkOutDate
            );
        } catch (RuntimeException e) {
            throw e;
        } catch (Exception e) {
            throw new RuntimeException("Error calculating checkout summary: " + e.getMessage(), e);
        }
    }

    /**
     * Process checkout with billing calculation
     */
    @Transactional
    public BookingResponseDTO processCheckout(Long id) {
        try {
            Booking booking = bookingRepository.findById(id)
                    .orElseThrow(() -> new RuntimeException("Booking not found with id: " + id));

            if (booking.getStatus() != Booking.BookingStatus.CHECKED_IN) {
                throw new RuntimeException("Cannot check out: Guest is not checked in");
            }

            // Get checkout summary for billing calculation
            CheckoutSummaryDTO checkoutSummary = getCheckoutSummary(id);
            
            // Update booking with final amount
            booking.setTotalAmount(checkoutSummary.getGrandTotal());
            
            // Update status to checked out
            booking.setStatus(Booking.BookingStatus.CHECKED_OUT);
            
            Booking updatedBooking = bookingRepository.save(booking);
            return new BookingResponseDTO(updatedBooking);
        } catch (RuntimeException e) {
            throw e;
        } catch (Exception e) {
            throw new RuntimeException("Error processing checkout: " + e.getMessage(), e);
        }
    }

    /**
     * Get today's check-ins for dashboard
     */
    public List<CheckInSummaryDTO> getTodayCheckIns() {
        try {
            List<Booking> bookings = bookingRepository.findTodayCheckIns(LocalDate.now());
            return bookings.stream()
                    .map(booking -> {
                        String guestName = booking.getFirstName() + " " + booking.getLastName();
                        String roomNumber = booking.getRoom().getRoomNumber();
                        LocalDate checkInDate = booking.getCheckInDate();
                        LocalDate checkOutDate = booking.getCheckOutDate();
                        int daysRemaining = (int) ChronoUnit.DAYS.between(LocalDate.now(), checkOutDate);
                        
                        return new CheckInSummaryDTO(
                            booking.getId(),
                            guestName,
                            roomNumber,
                            checkInDate,
                            checkOutDate,
                            Math.max(0, daysRemaining)
                        );
                    })
                    .collect(Collectors.toList());
        } catch (Exception e) {
            throw new RuntimeException("Error fetching today's check-ins: " + e.getMessage(), e);
        }
    }

    /**
     * Get today's check-outs for dashboard
     */
    public List<CheckInSummaryDTO> getTodayCheckOuts() {
        try {
            List<Booking> bookings = bookingRepository.findTodayCheckOuts(LocalDate.now());
            return bookings.stream()
                    .map(booking -> {
                        String guestName = booking.getFirstName() + " " + booking.getLastName();
                        String roomNumber = booking.getRoom().getRoomNumber();
                        LocalDate checkInDate = booking.getCheckInDate();
                        LocalDate checkOutDate = booking.getCheckOutDate();
                        int daysRemaining = 0; // Already checking out today
                        
                        return new CheckInSummaryDTO(
                            booking.getId(),
                            guestName,
                            roomNumber,
                            checkInDate,
                            checkOutDate,
                            daysRemaining
                        );
                    })
                    .collect(Collectors.toList());
        } catch (Exception e) {
            throw new RuntimeException("Error fetching today's check-outs: " + e.getMessage(), e);
        }
    }
}